pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,188 @@
+// rotate and delete old logs in file based on TTL
+func (l *ECSErrorLogger) ECSErrorReporter() {
+	defer l.ticker.Stop()
+	for {
+		select {
+		// when new error add in the channel, this case will be selected and write it into error log file
+		case e := <-errorLogChannel:
+			l.Write(e)
+		// this case will run every x mins (x is configurable) to delete the error which already expired, error log max age
+		// depends on STATUS_MESSAGE_TTL
+		case <-l.ticker.C:
+			l.processTimeout()
+		}
+	}
+}
+
+func (l *ECSErrorLogger) Write(newError zapcore.Entry) {
+	if l.errorFile == nil {
+		if err := l.openExistingOrNewErrorFile(); err != nil {
+			log.Printf("could not open error log file when write new error, err: %v", err)
+		}
+	}
+	newErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+		newError.Time, newError.Level, newError.Caller, newError.Message))
+
+	if len(newErrorLog) > l.ErrorFileMaxSize {
+		log.Printf("error size exceed the max size of error file")
+		return
+	}
+
+	err := l.errorFile.Truncate(0)
+	if err != nil {
+		log.Printf("could not truncate error log file when write new error, err: %v", err)
+	}
+	l.size = 0
+
+	errorNum := len(l.queue)
+	// iterate errors in queue, if the old error's message same as the new error's message, skip it,
+	// otherwise write it into the file and queue
+	for i := 0; i < errorNum; i++ {
+		e := l.queue[0]
+		if e.Message != newError.Message {
+			n, err := l.errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				log.Printf("could not write error into error log file, err: %v", err)
+			}
+			l.size += n
+			l.queue = append(l.queue, e)
+		}
+		l.queue = l.queue[1:]
+	}
+	// delete the old errors if old errors plus new error exceed the max size
+	if l.size+len(newErrorLog) > l.ErrorFileMaxSize {
+		l.rotate(newErrorLog)
+	}
+
+	l.queue = append(l.queue, newError)
+	n, err := l.errorFile.Write(newErrorLog)
+	if err != nil {
+		log.Printf("could not write new error into error log file, err: %v", err)
+	}
+	l.size += n
+}
+
+// rotate func could delete the old logs to make sure error log file less than 1 KB
+func (l *ECSErrorLogger) rotate(newErrorLog []byte) {
+	file, err := os.OpenFile(l.ErrorFilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+	if err != nil {
+		log.Printf("could not open error log file during rotation, err: %v", err)
+	}
+
+	err = file.Truncate(0)
+	if err != nil {
+		log.Printf("could not truncate error log file during rotation, err: %v", err)
+	}
+
+	errorNum := len(l.queue)
+	for i := 0; i < errorNum; i++ {
+		currentError := l.queue[0]
+		currentErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+			currentError.Time, currentError.Level, currentError.Caller, currentError.Message))
+		if l.size+len(newErrorLog) > l.ErrorFileMaxSize {
+			l.size -= len(currentErrorLog)
```

Comment:
I think the logic here is wrong...

Reply:
I think there is copy paste mistake here with `newErrorLog` being used.

Reply:
why there is a logic issue? for example we iterate 3 errors in queue, when `i = 0`, we found the `l.size+len(newErrorLog)` exceed the max size, we just minus the first error length from `l.size` and don't write it into the file. when `i=1` and `i=2`, we found that now `l.size+len(newErrorLog)` smaller than max size, so we still write the rest 2 errors into the file.
is this logic correct?

Reply:
I see. Ok I think your logic is technically correct, but I think its a bit confusing and not ideal.

It'd be simpler and easier to read if the logic was:

```
append new error log to end of queue

while size of queue is greater than max size:
     remove first log from queue (oldest log)
    
```